src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java

```diff
@@ -54,7 +54,18 @@ public TreeMaker() {}
     TypeMaker typeMaker = new TypeMaker(this);
 
     private <T extends Declaration> T checkCache(Cursor c, Class<T> clazz, Supplier<Declaration> factory) {
-        return clazz.cast(treeCache.computeIfAbsent(c, cx->factory.get()));
+        // The supplier function may lead to adding some other type, which will cause CME using computeIfAbsent.
+        // This implementation relax the constraint a bit only check for same key
```

Comment:
Nice workaround - basically only throw if we see that we're about to add something that was already there,